From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: gellert <gellert.hegyi@around.co>
Date: Thu, 12 Aug 2021 19:27:15 +0200
Subject: adds option to hide show cursor and change exclusion fps in desktop
 screensharing

This patch adds the ability to the desktop capturerer to hide and show cursor and change exclusion fps in desktop screensharing.

diff --git a/modules/desktop_capture/desktop_and_cursor_composer.cc b/modules/desktop_capture/desktop_and_cursor_composer.cc
index 400cf19ed863430f4f8499978b71236a6ad045e7..fb1afcd37924f6e3b0960e580135e79f88c42900 100644
--- a/modules/desktop_capture/desktop_and_cursor_composer.cc
+++ b/modules/desktop_capture/desktop_and_cursor_composer.cc
@@ -69,7 +69,8 @@ class DesktopFrameWithCursor : public DesktopFrame {
                          const MouseCursor& cursor,
                          const DesktopVector& position,
                          const DesktopRect& previous_cursor_rect,
-                         bool cursor_changed);
+                         bool cursor_changed,
+                         bool show_cursor);
   ~DesktopFrameWithCursor() override;
 
   DesktopRect cursor_rect() const { return cursor_rect_; }
@@ -89,7 +90,8 @@ DesktopFrameWithCursor::DesktopFrameWithCursor(
     const MouseCursor& cursor,
     const DesktopVector& position,
     const DesktopRect& previous_cursor_rect,
-    bool cursor_changed)
+    bool cursor_changed,
+    bool show_cursor)
     : DesktopFrame(frame->size(),
                    frame->stride(),
                    frame->data(),
@@ -119,16 +121,18 @@ DesktopFrameWithCursor::DesktopFrameWithCursor(
   restore_frame_->CopyPixelsFrom(*this, cursor_rect_.top_left(),
                                  DesktopRect::MakeSize(restore_frame_->size()));
 
-  // Blit the cursor.
-  uint8_t* cursor_rect_data =
-      reinterpret_cast<uint8_t*>(data()) + cursor_rect_.top() * stride() +
-      cursor_rect_.left() * DesktopFrame::kBytesPerPixel;
-  DesktopVector origin_shift = cursor_rect_.top_left().subtract(cursor_origin);
-  AlphaBlend(cursor_rect_data, stride(),
-             cursor.image()->data() +
-                 origin_shift.y() * cursor.image()->stride() +
-                 origin_shift.x() * DesktopFrame::kBytesPerPixel,
-             cursor.image()->stride(), cursor_rect_.size());
+  if (show_cursor) {
+    // Blit the cursor.
+    uint8_t* cursor_rect_data =
+        reinterpret_cast<uint8_t*>(data()) + cursor_rect_.top() * stride() +
+        cursor_rect_.left() * DesktopFrame::kBytesPerPixel;
+    DesktopVector origin_shift = cursor_rect_.top_left().subtract(cursor_origin);
+    AlphaBlend(cursor_rect_data, stride(),
+              cursor.image()->data() +
+                  origin_shift.y() * cursor.image()->stride() +
+                  origin_shift.x() * DesktopFrame::kBytesPerPixel,
+              cursor.image()->stride(), cursor_rect_.size());
+  }
 }
 
 DesktopFrameWithCursor::~DesktopFrameWithCursor() {
@@ -207,6 +211,8 @@ bool DesktopAndCursorComposer::IsOccluded(const DesktopVector& pos) {
 void DesktopAndCursorComposer::OnCaptureResult(
     DesktopCapturer::Result result,
     std::unique_ptr<DesktopFrame> frame) {
+  desktop_capturer_->exclusion_fps_ = exclusion_fps_;
+
   if (frame && cursor_) {
     if (!frame->may_contain_cursor() &&
         frame->rect().Contains(cursor_position_) &&
@@ -226,7 +232,7 @@ void DesktopAndCursorComposer::OnCaptureResult(
 #endif
       auto frame_with_cursor = std::make_unique<DesktopFrameWithCursor>(
           std::move(frame), *cursor_, relative_position, previous_cursor_rect_,
-          cursor_changed_);
+          cursor_changed_, show_cursor_);
       previous_cursor_rect_ = frame_with_cursor->cursor_rect();
       cursor_changed_ = false;
       frame = std::move(frame_with_cursor);
diff --git a/modules/desktop_capture/desktop_capturer.h b/modules/desktop_capture/desktop_capturer.h
index 6617c69e351bc7fd1e22cb8bc53c8b19dbf4c0f9..be9fbf47da5ab2c212bad639fb1e6343e39205d1 100644
--- a/modules/desktop_capture/desktop_capturer.h
+++ b/modules/desktop_capture/desktop_capturer.h
@@ -81,6 +81,8 @@ class RTC_EXPORT DesktopCapturer {
   // valid until capturer is destroyed.
   virtual void Start(Callback* callback) = 0;
   bool should_exclude_ = false;
+  bool show_cursor_ = true;
+  float exclusion_fps_ = 5.0f;
 
   // Sets SharedMemoryFactory that will be used to create buffers for the
   // captured frames. The factory can be invoked on a thread other than the one
diff --git a/modules/desktop_capture/mac/screen_capturer_mac.mm b/modules/desktop_capture/mac/screen_capturer_mac.mm
index 7a890c48ae91c139e68f7be292b316ae2e0e2363..c165b57195c6e066b935d804176f4a6ce6f95b99 100644
--- a/modules/desktop_capture/mac/screen_capturer_mac.mm
+++ b/modules/desktop_capture/mac/screen_capturer_mac.mm
@@ -389,8 +389,7 @@ DesktopRect GetExcludedWindowPixelBounds(CGWindowID window, float dip_to_pixel_s
     copy_region.Translate(-display_bounds.left(), -display_bounds.top());
 
     if (should_exclude_) {
-      float window_exclude_capture_fps = 5.0f;
-      float delta = 1000.0f / window_exclude_capture_fps;
+      float delta = 1000.0f / exclusion_fps_;
       if (last_exclude_capture_ms_ == 0 || rtc::TimeSince(last_exclude_capture_ms_) > delta) {
         last_exclude_capture_ms_ = rtc::SystemTimeMillis();
         excluded_images_.clear();
